Generalize Cosmos address to alternative pubkey hash, merge evmos

catenocrypt opened the pull request:
## Description

Generalize Cosmos address to support alternative pubkey hash, from config; merge native Evmos implementation with Cosmos.  NativeEvmos thus does not need a custom blockchain type, nor Entry, Address implementation.

## How to test

All unit tests, esp. Cosmos-based chains.

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

* New feature (non-breaking change which adds functionality)

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

hewigovens commented on registry.json:
```diff
@@ -2405,6 +2412,7 @@
     "curve": "secp256k1",
     "publicKeyType": "secp256k1Extended",
     "hrp": "evmos",
+    "publicKeyHasher": "keccak256",
```
This needs to add to all EVM items, not just Evmos

catenocrypt replied:
Ethereum implementation does not use publicKeyHasher field, now only Cosmos implementation was extended to use this field.  Other than Cosmos, only Bitcoin uses it.

hewigovens replied:
Yes, but this issue is existing for a while: developer might call `TW::publicKeyHasher(TWCoinTypeEthereum)` and get the default value `sha256ripemd`, so I think it's a good time to fix it

catenocrypt replied:
Sure, done.

hewigovens commented:
It seems that `publicKeyHasher` confuses us, being used as extended public key hasher and public key hasher in address, let's use a separate field in registry json? and keep `publicKeyHasher` as it is.

catenocrypt commented:
Suggestion:
- leave `publicKeyHasher` as it is, used in HDWallet xpub derivation 
- introduce `addressHasher`, used in hashing address, e.g. in Cosmos.  For Eth chains, this can be set to keccak.

hewigovens commented on src/Hash.cpp:
```diff
@@ -19,6 +19,29 @@
+        case Hash::HashFunc_sha3_512: return Hash::sha3_512;
+        case Hash::HashFunc_ripemd: return Hash::ripemd;
+        case Hash::HashFunc_blake256: return Hash::blake256;
+        //case Hash::HashFunc_blake2b: return static_cast<Hash::HasherSimpleType>(Hash::blake2b);
```
```suggestion
```

hewigovens commented on src/Hash.h:
```diff
@@ -12,10 +12,37 @@
+    HashFunc_sha3_512, // version 3 SHA512
+    HashFunc_ripemd, // RIPEMD160
+    HashFunc_blake256, // Blake256
+    //HashFunc_blake2b, // Blake2b
```
```suggestion
```

hewigovens reviewed (approved):
looks good, linux CI is red

catenocrypt commented:
I have added a few more Hash tests, to restore coverage (Linux CI failed because of coverage).
I also added documentation on fields in `registry.json`:

https://github.com/trustwallet/wallet-core/blob/ar/evmos-cosmos/docs/registry-fields.md

@hewigovens , please re-review.
